Fix inconsistent behaviour for `last(::Zip)` with differing iterator size types

jakewilliami opened the pull request:
This PR fixes two bugs relating to `last(::Zip)`, closing #58922.

First: there is inconsistent behaviour for `last(::Zip)` with differing iterator size types (a mixture of known-size and not) are passed in the `Zip`, causing a `MethodError`.

`last(::Zip)` works with finite zipped iterators of different lengths, but fails when one of them is an infinite iterator.  In the case where there are a mixture of known-length (finite) and infinite iterators in the `Zip`, we can know their length statically.  (I call these `Zip`s "finite-guarded," because the `Zip` is finite due to its finite component.)

The desired behaviour is to match the behaviour of `last` with finite iterators of different lengths.

Second: while standardising this behaviour, [a bug in `last(::Zip)` for `OffsetArray`s](https://github.com/JuliaLang/julia/issues/58922#issuecomment-3045850520) is fixed.  The issue was subtle and didn't error, but produced the wrong answer (noticed by @adienes).

We also add an explicit `ArgumentError` when `last` is called on a `Zip` whose size is unknown (good call by @Seelengrab).  Previously this was a `MethodError` for `lastindex`, used by the [previous implementation of `last(::Zip)`](https://github.com/JuliaLang/julia/blob/80f7db8e51b2ba1dd21e913611c23a6d5b75ecab/base/iterators.jl#L476).

This PR does **not** give support to any [functionality involving iterators of unknown size](https://github.com/JuliaLang/julia/pull/59217/files#discussion_r2256315847).  This may be done in future.

jakewilliami committed:
Add tests for asymmetric finite-guarded zips

Add tests for zips where one of the iterators are not bounded but the
zip iterator is bounded by the finite one.

Addresses #58922

@adienes had a good idea to test against OffsetArrays

jakewilliami committed:
Support `last` for finite-guarded zip iterators

Inconsistent behaviour for `last(::Zip)` with differing iterator size
types causes `MethodError`.  `last` works with finite zipped iterators
of different length, but fails when one of them is an infinite
iterator.  The desired behaviour is to match the behaviour of `last`
with finite iterators of different lengths.

Closes #58922

adienes commented on test/iterators.jl:
```diff
@@ -1206,3 +1204,18 @@ end
+    @test last(zip(1:3, Iterators.cycle(('x', 'y')))) == (3, 'x')
+    @test last(zip(1:3, Iterators.repeated('x'))) == (3, 'x')
+
+    # TODO: test for other array types
```
```
isdefined(Main, :OffsetArrays) || @eval Main include("testhelpers/OffsetArrays.jl")
using .Main.OffsetArrays
```

is a pattern used in many test suites to get `OffsetArrays` in Base tests (put at the top

jakewilliami replied:
Resolved in 86d2f58 (rebased to keep history clean)

adienes commented on base/iterators.jl:
```diff
@@ -1700,6 +1701,10 @@ function _nth(::IteratorSize, itr, n::Integer)
     y[1]
 end
+
+# issue #58922
```
```suggestion
```

this specialization is reasonable to add no matter the motivating issue :)

jakewilliami replied:
Resolved in 88639c0 (rebased to keep history clean)

jakewilliami commented on base/iterators.jl:
```diff
@@ -1700,6 +1701,10 @@ function _nth(::IteratorSize, itr, n::Integer)
 end
+
+# issue #58922
+_nth(::Union{HasShape, HasLength}, z::Zip, n::Integer) = map(nth(n), z.is)
```
Resolved in 88639c0 (rebased to keep history clean)

jakewilliami commented:
Thanks for the edits @adienes!  Re-committed them via `rebase` to keep the git history clean :slightly_smiling_face:

adienes commented:
seems like a good change to me

another pair of eyes is always good; maybe @Seelengrab or @jakobnissen are interested in giving feedback?

Seelengrab commented on base/iterators.jl:
```diff
@@ -1700,6 +1701,9 @@ function _nth(::IteratorSize, itr, n::Integer)
     y[1]
 end
+
+_nth(::Union{HasShape, HasLength}, z::Zip, n::Integer) = Base.map(nth(n), z.is)
```
What about `SizeUnknown`?

adienes replied:
it will hit the generic fallback above that just calls `iterate` n times

jakewilliami replied:
Great catch.  I didn't consider that case.  This does indeed fail on iterators of unknown size.  For example,
```julia
Iterators.filter(x -> x > 0, -5:5)
```
As humans, it's obvious that this is synonymous to `1:5`, but it doesn't implement `length`.

We might need use `_zip_lengths_finite_equal` (we explored using internal `Iterator` length functions in #58922 but initially decided against it).  Do you think this is sufficient (CC @adienes)?
```julia
function last(z::Zip)
    n = last(_zip_lengths_finite_equal(z.is))
    return nth(z, n)
end
```
Is it worth keeping the `length` implementation for iterators whose size is known?

jakewilliami replied:
I'm not sure how `last` can know the length of the `Zip` if all constituent iterators have `SizeUnknown`.  Do we explicitly error in this case?

`nth` supports iterators whose size are unknown.  It would be nice to support it in `last` but I don't see any efficient way to get the length.

jakewilliami replied:
`_zip_lengths_finite_equal` seems to handle the case where at least one of the constituent iterators have a known length, which our current implementation does not do:
```julia-repl
julia> z = zip(1:3, Iterators.filter(x -> x > 0, -5:5))
zip(1:3, Base.Iterators.Filter{var"#2#3", UnitRange{Int64}}(var"#2#3"(), -5:5))

julia> length(z)
ERROR: MethodError: no method matching length(::Base.Iterators.Filter{var"#2#3", UnitRange{Int64}})

julia> Iterators._zip_lengths_finite_equal(z.is)
(false, 3)
```

LilithHafner replied:
```suggestion
_nth(::IteratorSize, z::Zip, n::Integer) = Base.map(nth(n), z.is)
```
and defer to the component iterators.

jakewilliami replied:
Implemented in da76f8ce92 :slightly_smiling_face:

jakewilliami committed:
Explicit throw on `last(::Zip)` when size unknown

This should be explicitly handled (astutely noted by @Seelengrab).

Addresses #58922

jakewilliami commented:
Interesting, I don't think I knew that `last` asserts that it's $`O(1)`$ (or perhaps I forgot).  Nice catch.  It's a promise that was [made many years ago](https://github.com/JuliaLang/julia/commit/8db30c60afbb97a019ef9153ed340b2aee68337e#diff-bb45335438eec1f8480091a78cbe5b806d9948c19b1e2efc8d0ad8b69a3cf75c) by @JeffBezanson.

I don't think we should create an implementation of `last` that isn't $`O(1)`$, but I do think that there are valid issues addressed in #58922.  I will have a think as well.  At the very least, we should explicitly error in the case where we would be using `nth`, and (importantly) fix the bug in `last` on zipped `OffsetArray`s.

Seelengrab commented:
Is that promise still in effect? The linked doc seems to be exclusively about collections that are indexable in O(1), which is obviously impossible for general iterators. I don't see an issue with extending that.

Seelengrab commented:
Let's put that up to triage then, to get some more eyes on this & make a decision about whether it's ok for `last` to be something other than $O(1)$.

jakewilliami commented:
What of the time complexity of this implementation of [taking the last $`n`$ elements of an iterator](https://github.com/JuliaLang/julia/blob/46c2a5c7e1f970e83b408b6ddcba49aaa31d8329/base/abstractarray.jl#L551)?
```julia
last(itr, n::Integer) = reverse!(collect(Iterators.take(Iterators.reverse(itr), n)))
```

LilithHafner commented on base/iterators.jl:
```diff
@@ -473,7 +473,12 @@ zip_iteratoreltype() = HasEltype()
+function last(z::Zip)
+    IteratorSize(z) == SizeUnknown() &&
+        throw(ArgumentError("Cannot get last element of zipped iterators of undefined lengths"))
+    return nth(z, length(z))
```
Changing 
`getindex.(z.is, minimum(Base.map(lastindex, z.is)))`
to
`nth(z, length(z))`
is a clear win and an easy merge. Thanks!

LilithHafner commented on base/iterators.jl:
```diff
@@ -473,7 +473,12 @@ zip_iteratoreltype() = HasEltype()
 
-last(z::Zip) = getindex.(z.is, minimum(Base.map(lastindex, z.is)))
+function last(z::Zip)
+    IteratorSize(z) == SizeUnknown() &&
+        throw(ArgumentError("Cannot get last element of zipped iterators of undefined lengths"))
```
This is breaking. It might be only slightly breaking but even then it would need to be well justified. (`SizeUnknown` does not guarantee that length will throw)

jakewilliami committed:
Dispatch internal nth method on IteratorSize

Rather than implenenting internal nth method for HasShape and
HasLength, dispatch generally on IteratorSize and defer to the
component iterators.

Suggestion by @LilithHafner

jakewilliami committed:
Remove throw for iterator of unknown size in last(::Zip)

Although it may be useful to through an ArgumentError rather than a
MethodError if we can't get the length of an iterator whose size is
unknown, SizeUnknown does not guarantee that length will throw.
Therefore, we remove this check and defer to a MethodError if the
iterator does not implement length.

Suggestion by @LilithHafner
